tools/envoy_collect.py

```diff
@@ -0,0 +1,195 @@
+  - Support other modes of data collection (e.g. snapshotting on SIGUSR,
+    periodic).
+  - Consider handling other signals.
+  - bz2 compress tar ball.
+"""
+
+import argparse
+import ctypes
+import ctypes.util
+import datetime
+import json
+import os
+import pipes
+import shutil
+import signal
+import subprocess as sp
+import sys
+import tarfile
+import tempfile
+import urllib2
+
+ENVOY_PATH = os.getenv(
+    'ENVOY_PATH',
+    'bazel-out/local-fastbuild/genfiles/source/exe/envoy-static.stamped')
+PERF_PATH = os.getenv('PERF_PATH', 'perf')
+
+PR_SET_PDEATHSIG = 1  # See prtcl(2).
+
+DUMP_HANDLERS = ['clusters', 'listeners', 'server_info', 'stats']
+
+
+def FetchUrl(url):
+  return urllib2.urlopen(url).read()
+
+
+def ModifyEnvoyconfig(config_path, output_directory):
+  # Load original Envoy config.
+  with open(config_path, 'r') as f:
+    envoy_config = json.loads(f.read())
+
+  # Add unconditional access logs for all listeners.
+  access_log_paths = []
+  for n, listener in enumerate(envoy_config['listeners']):
+    for network_filter in listener['filters']:
+      if network_filter['name'] == 'http_connection_manager':
+        config = network_filter['config']
+        access_log_path = os.path.join(output_directory, 'access_%d.log' % n)
+        access_log_config = {'path': access_log_path}
+        if 'access_log' in config:
+          config['access_log'].append(access_log_config)
+        else:
+          config['access_log'] = [access_log_config]
+        access_log_paths.append(access_log_path)
+
+  # Write out modified Envoy config.
+  modified_envoy_config_path = os.path.join(output_directory, 'config.json')
+  with open(modified_envoy_config_path, 'w') as f:
+    f.write(json.dumps(envoy_config, indent=2))
+
+  return modified_envoy_config_path, access_log_paths
+
+
+def EnvoyCollect(parse_result, unknown_args):
+  envoy_tmpdir = tempfile.mkdtemp()
+  manifest = []
+  return_code = 1
+  try:
+    # Setup Envoy config and determine the paths of the files we're going to
+    # generate.
+    modified_envoy_config_path, access_log_paths = ModifyEnvoyconfig(
+        parse_result.config_path, envoy_tmpdir)
+    dump_handlers_paths = {
+        h: os.path.join(envoy_tmpdir, '%s.txt' % h)
+        for h in DUMP_HANDLERS
+    }
+    perf_data_path = os.path.join(envoy_tmpdir, 'perf.data')
+    envoy_log_path = os.path.join(envoy_tmpdir, 'envoy.log')
+    # The manifest of files that will be placed in the output .tar.
+    manifest = access_log_paths + dump_handlers_paths.values() + [
+        modified_envoy_config_path, perf_data_path, envoy_log_path
+    ]
+
+    # This is where we will find out where the admin endpoint is listening.
+    admin_address_path = os.path.join(envoy_tmpdir, 'admin_address.txt')
+
+    # This is how we will invoke the wrapped envoy.
+    # TODO(htuch): Only run under perf when we want a profile, not during debug.
+    envoy_shcmd = ' '.join(
+        map(pipes.quote, [
```

Comment:
Thanks for the feedback on py3 in general. I'll make a best effort to support this, unless it becomes too difficult.

I'll keep `pipes.quote` if it doesn't break py3, since 2.7.6 doesn't support `shlex.quote` yet.